test/realm_logger_test.dart

```diff
@@ -0,0 +1,337 @@
+      ..onRecord.listen((event) {
+        events.add(event);
+      });
+  } else {
+    if (logLevel != null) {
+      Realm.logger.level = logLevel;
+    }
+    Realm.logger.onRecord.listen((event) {
+      events.add(event);
+    });
+  }
+
+  try {
+    await action();
+    await completer.future.timeout(Duration(seconds: 2), onTimeout: () => throw Exception("Stop waiting for logs"));
+  } catch (error) {
+    completer.complete();
+  }
+
+  return events;
+}
+
+Future<List<LogRecord>> _attachToLoggerAfterAction(String isolateName, FutureOr<void> Function() action, {Level? logLevel, bool setNewLogger = false}) async {
+  final completer = Completer<void>();
+  await Future<void>.delayed(Duration(milliseconds: 200));
+  List<LogRecord> events = [];
+  try {
+    await action();
+    if (setNewLogger) {
+      Realm.logger = Logger.detached(generateRandomString(10))
+        ..level = logLevel ?? RealmLogLevel.info
+        ..onRecord.listen((event) {
+          events.add(event);
+        });
+    } else {
+      if (logLevel != null) {
+        Realm.logger.level = logLevel;
+      }
+      Realm.logger.onRecord.listen((event) {
+        events.add(event);
+      });
+    }
+
+    await completer.future.timeout(Duration(seconds: 2), onTimeout: () => throw Exception("Stop waiting for logs"));
+  } catch (error) {
+    completer.complete();
+  }
+  return events;
+}
+
+Future<void> _simulateLogs(String message) async {
+  RealmInternal.log(RealmLogLevel.trace, message);
+  RealmInternal.log(RealmLogLevel.debug, message);
+  RealmInternal.log(RealmLogLevel.detail, message);
+  RealmInternal.log(RealmLogLevel.info, message);
+  RealmInternal.log(RealmLogLevel.fatal, message);
+  RealmInternal.log(RealmLogLevel.error, message);
+  RealmInternal.log(RealmLogLevel.warn, message);
+}
+
+Map<Level, int> _matchedMessagesCountPerLevel(List<LogRecord>? logRecords, String specificMessage) {
+  Map<Level, int> matchedMessages = {};
+  for (Level level in Level.LEVELS) {
+    final matches = logRecords?.where((element) => element.level == level && element.message.contains(specificMessage));
+    matchedMessages.addEntries([MapEntry(level, matches?.length ?? 0)]);
+  }
+  return matchedMessages;
+}
+
+void _expectLogMessages(
+  Map<Level, int> actualMessages, {
+  List<Level> notExpectedMessagesFromLevels = const [],
+  List<Level> expectedMessagesFromLevels = const [],
+}) {
+  // Check count of various levels
+
+  for (final level in expectedMessagesFromLevels) {
+    final int count = actualMessages[level] ?? 0;
+    expect(count, greaterThan(0), reason: 'There is no messages from $level');
+  }
+
+  for (final level in notExpectedMessagesFromLevels) {
+    final count = actualMessages[level] ?? 0;
+    expect(count, 0, reason: 'Messages from level $level are not expected:\n  found ${actualMessages[level] ?? 0}');
+  }
+}
+
+void _expectErrorLogs(Map<String, Map<Level, int>> logMessages, String expectedName) {
```

Comment:
Could we remove the underscore in the common functions in each tests library. It will make it a bit more readable.